src/number_theory/modular.lean

```diff
@@ -363,6 +393,269 @@ begin
+  /- Any point `w∈𝒟ᵒ` has imaginary part at least `sqrt (3/4)` -/
+  have ImGeInD : ∀ (w : ℍ), w ∈ 𝒟ᵒ → 3/4 < (w.im)^2,
+  { intros w hw,
+    have : 1 < w.re * w.re + w.im * w.im := by simpa [complex.norm_sq_apply] using hw.1,
+    have := hw.2,
+    cases abs_cases w.re; nlinarith, },
+  /- The next argument is simply that `c^2 y^2 ≤ |c z + d|^2`. -/
+  have czPdGecy : (g 1 0 : ℝ)^2 * (z.im)^2 ≤ norm_sq (denom g z) :=
+    calc
+    (g 1 0 : ℝ)^2 * (z.im)^2 ≤ (g 1 0 : ℝ)^2 * (z.im)^2 + (g 1 0 * z.re + g 1 1)^2 : by nlinarith
+    ... = norm_sq (denom g z) : by simp [norm_sq]; ring,
+  have zIm : (3 : ℝ) / 4 < (z.im)^2 := ImGeInD _ hz,
+  /- This is the main calculation:
+  `sqrt 3 / 2 < Im(g•z) = Im(z)/|cz+d|^2 ≤ y/(c^2 y^2) < 2/(c^2 sqrt 3)`
+  -/
+  calc
+  (3 : ℝ) / 4 < ((g • z).im) ^ 2 : ImGeInD _ hg
+  ... = (z.im) ^ 2 / (norm_sq (denom g z)) ^ 2 : _
+  ... ≤ (1 : ℝ) / ((g 1 0) ^ 4 * (z.im) ^ 2) : _
+  ... < (4 : ℝ) / (3 * (g 1 0) ^ 4) : _,
+  { convert congr_arg (λ (x:ℝ), x ^ 2) (im_smul_eq_div_norm_sq g z) using 1,
+    exact (div_pow _ _ 2).symm, },
+  { rw div_le_div_iff,
+    convert pow_le_pow_of_le_left _ czPdGecy 2 using 1;
+    ring_nf,
+    { nlinarith, },
+    { exact pow_two_pos_of_ne_zero _ (norm_sq_denom_ne_zero g z), },
+    { nlinarith, }, },
+  { rw div_lt_div_iff,
+    repeat {nlinarith}, },
+end
+
+
+/-- Knowing that `3/4<4/(3c^4)` from `ineq_1`, and `c≠0`, we conclude that `c=1` or `c=-1`. -/
+lemma ineq_2 (c : ℤ) (hc₁ : (3 : ℝ)/4 < 4/ (3* c^4)) (hc₂ : c ≠ 0) : c = 1 ∨ c = -1 :=
+begin
+  rcases le_or_gt (|c|) 1 with h | (h : 2 ≤ |c|),
+  { -- case |c| ≤ 1
+    obtain ⟨h1c, hc1⟩ : -1 ≤ c ∧ c ≤ 1 := abs_le.mp h,
+    interval_cases c; tauto },
+  { -- case 2 ≤ |c|
+    exfalso,
+    have : 2^4 ≤ c^4,
+    { refine pow_four_le_pow_four _,
+      convert h using 1, },
+    have : (2:ℝ)^4 ≤ c^4,
+    { norm_cast,
+      convert this using 1, },
+    have := (div_lt_div_iff _ _).mp hc₁,
+    repeat {linarith}, },
+end
+
+/-- Definition: `T_pow` is the matrix `T` raised to the power `n:ℤ`. -/
+def T_pow (n : ℤ) : SL(2,ℤ) := ⟨ ![![1, n],![0,1]],
+begin
+  rw matrix.det_fin_two,
+  simp,
+end ⟩
+
+/- If c=1, then `g=[[1,a],[0,1]] * S * [[1,d],[0,1]]`. -/
+lemma g_eq_of_c_eq_one (g : SL(2,ℤ)) (hc : ↑ₘg 1 0 = 1) :
+  g = T_pow (g 0 0) * S * T_pow (g 1 1) :=
+begin
+  rw [T_pow, T_pow],
+  ext i,
+  fin_cases i; fin_cases j,
+  { simp [S, matrix.mul_apply, fin.sum_univ_succ] },
+  { have g_det : (1:ℤ) = ↑ₘg 0 0 * ↑ₘg 1 1 - 1 * ↑ₘg 0 1,
+    { convert det_fin_two ↑ₘg using 1,
+      { rw g.det_coe },
+      rw hc,
+      ring },
+    simp [S, matrix.mul_apply, fin.sum_univ_succ],
+    rw g_det,
+    simp, },
+  { simpa [S, matrix.mul_apply, fin.sum_univ_succ] using hc },
+  { simp [S, matrix.mul_apply, fin.sum_univ_succ], },
+end
+
+/-- Nontrivial lemma: if `|x|<1/2` and `n:ℤ`, then `2nx+n^2≥0`. (False for `n:ℝ`!) -/
+lemma _root_.int.non_neg_of_lt_half (n : ℤ) (x : ℝ) (hx : |x| < 1/2) : (0:ℝ) ≤ 2 * n * x + n * n :=
```

Comment:
How about if we start out with:
```lean
lemma cast_one_le_of_pos {n : ℤ} (hn : 0 < n) : (1 : ℝ) ≤ n :=
begin
  rw ← cast_one, 
  exact cast_le.mpr (add_one_le_of_lt hn),
end

lemma cast_le_neg_one_of_neg {n : ℤ} (hn : n < 0) : (n : ℝ) ≤ -1 :=
begin
  rw ← cast_one,
  exact cast_le.mpr (le_sub_one_of_lt hn),
end

lemma nneg_mul_add_sq_of_abs_le_one (n : ℤ) (x : ℝ) (hx : |x| ≤ 1) : (0 : ℝ) ≤ n * x + n * n :=
begin
  have hnx : 0 < n → 0 ≤ x + n := λ hn, by
  { convert add_le_add (neg_le_of_abs_le hx) (cast_one_le_of_pos hn),
    rw add_left_neg, },
  have hnx' : n < 0 → x + n ≤ 0 := λ hn, by
  { convert add_le_add (le_of_abs_le hx) (cast_le_neg_one_of_neg hn),
    rw add_right_neg, },
  rw [← mul_add, mul_nonneg_iff],
  rcases lt_trichotomy n 0 with h | rfl | h,
  { exact or.inr ⟨cast_nonpos.mpr h.le, hnx' h⟩, },
  { simp [le_total 0 x], },
  { exact or.inl ⟨cast_nonneg.mpr h.le, hnx h⟩, },
end
```
and obtain what you want from this?

Reply:
Thanks! Where would you suggest these lemmata belong?

Reply:
Presumably those generalize to types other than ℝ?

Reply:
Hmm the last one certainly requires `n` to be an integer, and we need `≤`... What do you have in mind?

Reply:
I'm talking about generalizing away the ` ℝ` not the `ℤ`

Reply:
I tried a bit but got stuck in a few places...?

Reply:
Not to mention: I'm not even sure this kind of lemma would be useful anywhere else...? Is it worth generalizing away from ℝ at this point? (Maybe someone will later come and find a need for it in another context, and can generalize it then?..)

Reply:
The way I'd recommend doing this is to keep your existing proof, but replace ` ℝ` with `R` and add all the typeclasses it asks for

Reply:
It works now for linear_ordered_ring; where do you suggest these lemmata go? Thanks!

Reply:
I think the lemmas `cast_one_le_of_pos`, `cast_le_neg_one_of_neg`, `nneg_mul_add_sq_of_abs_le_one` could all go in `data/int/cast.lean` about here: https://github.com/leanprover-community/mathlib/blob/49e41eb5383e372352490c6236cb0293e4e84d59/src/data/int/cast.lean#L191

This will also place them into the `int` namespace which is appropriate.